consul/state/catalog.go

```diff
@@ -655,23 +681,51 @@ func (s *StateStore) NodeService(nodeName string, serviceID string) (uint64, *st
 }
 
 // NodeServices is used to query service registrations by node ID.
-func (s *StateStore) NodeServices(ws memdb.WatchSet, nodeName string) (uint64, *structs.NodeServices, error) {
+func (s *StateStore) NodeServices(ws memdb.WatchSet, nodeNameOrID string) (uint64, *structs.NodeServices, error) {
 	tx := s.db.Txn(false)
 	defer tx.Abort()
 
 	// Get the table index.
 	idx := maxIndexTxn(tx, "nodes", "services")
 
-	// Query the node
-	watchCh, n, err := tx.FirstWatch("nodes", "id", nodeName)
+	// Query the node by node name
+	watchCh, n, err := tx.FirstWatch("nodes", "id", nodeNameOrID)
 	if err != nil {
 		return 0, nil, fmt.Errorf("node lookup failed: %s", err)
 	}
-	ws.Add(watchCh)
-	if n == nil {
-		return 0, nil, nil
+
+	if n != nil {
+		ws.Add(watchCh)
+	} else {
+		if len(nodeNameOrID) < minUUIDLookupLen {
+			ws.Add(watchCh)
+			return 0, nil, nil
+		}
+
+		// Attempt to lookup the node by it's node ID
+		iter, err := tx.Get("nodes", "uuid_prefix", nodeNameOrID)
+		if err != nil {
+			return 0, nil, fmt.Errorf("node ID lookup failed: %s", err)
+		}
+		n = iter.Next()
+		if n == nil {
+			ws.Add(watchCh)
+			return 0, nil, nil
+		}
+
+		idWatchCh := iter.WatchCh()
+		if iter.Next() != nil {
```

Comment:
Also wrt the comment below - it's not that we didn't find anything - there were too many things.